test/integration/scheduler/framework_test.go

```diff
@@ -616,6 +667,198 @@ func TestUnreservePlugin(t *testing.T) {
+			Enabled: []schedulerconfig.Plugin{{Name: bindPlugin1.Name()}, {Name: bindPlugin2.Name()}},
+		},
+		PostBind: &schedulerconfig.PluginSet{
+			Enabled: []schedulerconfig.Plugin{{Name: postbindPlugin.Name()}},
+		},
+	}
+	// Set reserve and bind config for testing
+	pluginConfig := []schedulerconfig.PluginConfig{
+		{
+			Name: unreservePlugin.Name(),
+			Args: runtime.Unknown{},
+		},
+		{
+			Name: bindPlugin1.Name(),
+			Args: runtime.Unknown{},
+		},
+		{
+			Name: bindPlugin2.Name(),
+			Args: runtime.Unknown{},
+		},
+		{
+			Name: postbindPlugin.Name(),
+			Args: runtime.Unknown{},
+		},
+	}
+
+	// Create the master and the scheduler with the test plugin set.
+	context := initTestSchedulerWithOptions(t, testContext,
+		false, nil, registry, plugins, pluginConfig, false, time.Second)
+	defer cleanupTest(t, context)
+
+	cs := context.clientSet
+	// Add a few nodes.
+	_, err := createNodes(cs, "test-node", nil, 2)
+	if err != nil {
+		t.Fatalf("Cannot create nodes: %v", err)
+	}
+
+	tests := []struct {
+		bindPluginStatuses     []*framework.Status
+		expectBoundByScheduler bool   // true means this test case expecting scheduler would bind pods
+		expectBoundByPlugin    bool   // true means this test case expecting a plugin would bind pods
+		expectBindPluginName   string // expecting plugin name to bind pods
+		expectInvokeEvents     []pluginInvokeEvent
+	}{
+		// bind plugins skiped to bind the pod and scheduler bond the pod
+		{
+			bindPluginStatuses:     []*framework.Status{framework.NewStatus(framework.Skip, ""), framework.NewStatus(framework.Skip, "")},
+			expectBoundByScheduler: true,
+			expectInvokeEvents:     []pluginInvokeEvent{{pluginName: bindPlugin1.Name(), val: 1}, {pluginName: bindPlugin2.Name(), val: 1}, {pluginName: postbindPlugin.Name(), val: 1}},
+		},
+		// bindplugin2 succeeded to bind the pod
+		{
+			bindPluginStatuses:   []*framework.Status{framework.NewStatus(framework.Skip, ""), framework.NewStatus(framework.Success, "")},
+			expectBoundByPlugin:  true,
+			expectBindPluginName: bindPlugin2.Name(),
+			expectInvokeEvents:   []pluginInvokeEvent{{pluginName: bindPlugin1.Name(), val: 1}, {pluginName: bindPlugin2.Name(), val: 1}, {pluginName: postbindPlugin.Name(), val: 1}},
+		},
+		// bindplugin1 succeeded to bind the pod
+		{
+			bindPluginStatuses:   []*framework.Status{framework.NewStatus(framework.Success, ""), framework.NewStatus(framework.Success, "")},
+			expectBoundByPlugin:  true,
+			expectBindPluginName: bindPlugin1.Name(),
+			expectInvokeEvents:   []pluginInvokeEvent{{pluginName: bindPlugin1.Name(), val: 1}, {pluginName: postbindPlugin.Name(), val: 1}},
+		},
+		// bind plugin fails to bind the pod
+		{
+			bindPluginStatuses: []*framework.Status{framework.NewStatus(framework.Error, "failed to bind"), framework.NewStatus(framework.Success, "")},
+			expectInvokeEvents: []pluginInvokeEvent{{pluginName: bindPlugin1.Name(), val: 1}, {pluginName: unreservePlugin.Name(), val: 1}},
```

Comment:
How does this address the problem of the scheduler retrying the pod?

Reply:
Actually I'm not trying to stop scheduler from retrying the pod. I'm using chan and events to prove if bind plugin fails to bind a pod, unreserve plugin will be called instead of postbind plugin.

I wrote the following program to prove that it is race to check the relation between numUnreserveCalled and numBindCalled if binding failure. Running the program will print `x` as 1, 0 and even ***-1*** ... which proves it is impossible to check the result of `numBindCalled -numUnreserveCalled`.

```
func main() {
	var numBindCalled, numUnreserveCalled int
	go func() {
		for {
			// the scheduling goroutine
			time.Sleep(time.Millisecond*1)
			numBindCalled++
			numUnreserveCalled++
		}
	}()
	var x int
	for {
		// the test goroutine
		time.Sleep(time.Millisecond*1)
		x = numBindCalled - numUnreserveCalled
		fmt.Printf("x %d\n" , x)
	}
}
```
So I went through chan and events ways. Should I change this line to 

```
expectInvokeEvents: []pluginInvokeEvent{{pluginName: bindPlugin1.Name(), val: 1}, {pluginName: unreservePlugin.Name(), val: 1}, {pluginName: bindPlugin1.Name(), val: 2}, {pluginName: unreservePlugin.Name(), val: 2}},
```

to make it more clear?


By the way I think the `TestUnreservePlugin` case https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler/framework_test.go#L592-L594 is flaky too.

Reply:
I am not questioning the use of events and channels. I think that part makes sense. The problem is that you expect bind and unreserve to be called only once. If the scheduler retries the pod, aren't those plugins called more than once?

Reply:
> The problem is that you expect bind and unreserve to be called only once

My past test case expect this and I deleted it in the second commit of this PR.

```golang
-                       if err = wait.Poll(10*time.Millisecond, 30*time.Second, func() (done bool, err error) {
-                               return unresPlugin.numUnreserveCalled == 1, nil
-                       }); err != nil {
-                               t.Errorf("test #%v: Expected the unreserve plugin to be called once, was called %d times.", i, unresPlugin.numUnreserveCalled)
```

I now expect it will be called more than once and changed `bind plugin fails to bind the pod` case to expect a second term scheduling cycle. See the case expect 4 invoke events, 2 from bindPlugin1 and 2 from unreservePlugin.

```golang
		// bind plugin fails to bind the pod
		{
			bindPluginStatuses: []*framework.Status{framework.NewStatus(framework.Error, "failed to bind"), framework.NewStatus(framework.Success, "")},
			expectInvokeEvents: []pluginInvokeEvent{{pluginName: bindPlugin1.Name(), val: 1}, {pluginName: unreservePlugin.Name(), val: 1}, {pluginName: bindPlugin1.Name(), val: 2}, {pluginName: unreservePlugin.Name(), val: 2}},
		},
```